### PR TITLE
Handle csrf protection according to spec

### PR DESCRIPTION
**Pull request by ngoquang2708**
@red-avtovo Please ignore session-close and free-space related commits, The are in other PRs.
I will rebase this PR later when they are merged.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -561,25 +540,54 @@ impl TransClient {
...
-        );
-        let resp: reqwest::Response = rq.send().await?;
-        let rpc_response: RpcResponse<RS> = resp.json().await?;
-        info!("Response body: {:#?}", rpc_response);
-        Ok(rpc_response)
+        let mut remaining_retry = 5; // TODO Config
+        loop {
+            if remaining_retry <= 0 {
+                error!("Out of retries!")
+                //break Err { Box::new("")??? }; // FIXME
```
Comment (ngoquang2708) [inline src/lib.rs:551]: How could I return error here?

Comment (red-avtovo) [reply src/lib.rs:551]: Try to return an `Err(...)` with error object

File: src/lib.rs [hunk 2]
```diff
@@ -38,45 +44,19 @@ impl TransClient {
...
-            .header(CONTENT_TYPE, "application/json")
-    }
-
-    /// Performs session-get call and takes the x-transmission-session-id
-    /// header to perform calls, using it's value
-    ///
-    /// # Errors
-    ///
-    /// If response is impossible to unwrap then it will return an empty session_id
-    async fn get_session_id(&self) -> String {
```
Comment (red-avtovo) [inline src/lib.rs:63]: Why did you remove that?

File: src/lib.rs [hunk 3]
```diff
@@ -22,6 +24,8 @@ use types::{TorrentAddArgs, TorrentAdded};
 pub struct TransClient {
     url: String,
     auth: Option<BasicAuth>,
+    session_id: Option<String>,
+    client: Client,
```
Comment (red-avtovo) [inline src/lib.rs:28]: Is that really necessary to expose the client? Please try to avoid it

File: src/lib.rs [hunk 4]
```diff
@@ -111,7 +91,7 @@ impl TransClient {
     ///     Ok(())
     /// }
     /// ```
-    pub async fn session_get(&self) -> Result<RpcResponse<SessionGet>> {
+    pub async fn session_get(&mut self) -> Result<RpcResponse<SessionGet>> {
```
Comment (red-avtovo) [inline src/lib.rs:94]: Why should we allow to mutate the client within the instance? That looks dangerous

File: src/lib.rs [hunk 5]
```diff
@@ -147,10 +127,46 @@ impl TransClient {
...
+    ///     let response: Result<RpcResponse<SessionClose>> = client.session_close().await;
+    ///     match response {
+    ///         Ok(_) => println!("Yay!"),
+    ///         Err(_) => panic!("Oh no!")
+    ///     }
+    ///     println!("Rpc response is ok: {}", response?.is_ok());
+    ///     Ok(())
+    /// }
+    /// ```
+    pub async fn session_close(&mut self) -> Result<RpcResponse<SessionClose>> {
```
Comment (red-avtovo) [inline src/lib.rs:166]: docs and definition are not cohesive

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -41,6 +41,13 @@ jobs:
           args: --release --all-features
 
       - name: Cargo Test
-        uses:  actions-rs/cargo@v1
+        uses: actions-rs/toolchain@v1
+        with:
+          toolchain: stable
```
Comment (red-avtovo) [inline .github/workflows/ci.yml:46]: The Toolchain is already initialized

File: src/lib.rs (hunk 1 above)
Comment (ngoquang2708) [reply src/lib.rs:551]: So we have to define an error struct then?

File: src/lib.rs (hunk 2 above)
Comment (ngoquang2708) [reply src/lib.rs:63]: In Spec they said that we should just send a req, then check for 409(6?) Http response, then obtain session id from there because it may change. Beside that, I think we should catch that session id for later use, do not need to get it on every req.

File: src/lib.rs (hunk 3 above)
Comment (ngoquang2708) [reply src/lib.rs:28]: Aren't they private by default in Rust?

File: src/lib.rs (hunk 5 above)
Comment (ngoquang2708) [reply src/lib.rs:166]: Please ignore this

File: .github/workflows/ci.yml (hunk 6 above)
Comment (ngoquang2708) [reply .github/workflows/ci.yml:46]: Ignore this.

**Comment by red-avtovo** [comment]
@ngoquang2708, is there any work left?

**Comment by ngoquang2708** [comment]
@red-avtovo Still working on it :)

**Comment by red-avtovo** [comment]
Thanks for your update. Take your time. Much appreciated for your work

**Review by red-avtovo — APPROVED** [review]
Thank you very much for your hard work. Sorry for so late approval
